app/web/features/auth/useAuthStore.ts

```diff
@@ -29,10 +29,34 @@ export function usePersistedState<T>(
     },
     [key]
   );
+  return [_state, setState];
+}
+
+export function useSessionStorage<T>(
```

Comment:
Considering the implementation is more or less the same as `usePersistedState`... I feel like it might be worth adding a third optional argument for the hook instead so that
- we can swap between `window.localStorage`/`window.sessionStorage` as needed via this argument, defaulting to `localStorage` (because really, this seems to be the first use case where we needed to switch to `sessionStorage`)
- minimise duplication of the same code and,
- we don't break existing places that use the `usePersistedState` hook

Reply:
Yeah I hear you. Personally I tend to ignore duplicating code until it's repeated `>=3` times.

Would it be cool if I just add a `TODO` comment for this? I'm not sure how to handle the ssr part of this hook with a dynamic storage object.

Reply:
> how to handle the ssr part of this hook with a dynamic storage object.

Wouldn't it be the same as what's being checked in `saved` - i.e. it becomes something like:

```typescript
// storage is either `window.sessionStorage` or `window.localStorage`
const saved =
    typeof window !== "undefined" ? storage.getItem(key) : null;
```

I'd generally agree with your `>=3` heuristic too to avoid premature abstraction, but I think in this case, since the `localStorage` and `sessionStorage` API is exactly the same (they are both of type `Storage`), I see little risk of having to undo this even if we merge them together.

Reply:
If we add `storage = window.localStorage` as an argument to this function it's going to throw an error during ssr.

Were you thinking to do it with a flag? Something like this?

```typescript
export function usePersistedState<T>(
  key: string,
  defaultValue: T,
  isLocalStorage = true,
): [T | undefined, (value: T) => void, () => void] {
  //in ssr, window doesn't exist, just use default
  const saved = typeof window !== "undefined"
    ? isLocalStorage
      ? window.localStorage.getItem(key)
      : window.sessionStorage.getItem(key)
    : null

  const [_state, _setState] = useState<T | undefined>(
    saved !== null ? JSON.parse(saved) : defaultValue
  );
  const setState = useCallback(
    (value: T) => {
      if (value === undefined) {
        console.warn(`${key} can't be stored as undefined, casting to null.`);
      }
      const v = value === undefined ? null : value;
      const storage = isLocalStorage ? window.localStorage : window.sessionStorage
      storage.setItem(key, JSON.stringify(v));
      _setState(value);
    },
    [key, isLocalStorage]
  );
  const clearState = useCallback(() => {
    const storage = isLocalStorage ? window.localStorage : window.sessionStorage
    storage.removeItem(key);
    _setState(undefined);
  }, [key, isLocalStorage]);
  return [_state, setState, clearState];
}

export function useSessionStorage<T>(
  key: string,
  defaultValue: T,
): [T | undefined, (value: T) => void, () => void] {
  return usePersistedState(key, defaultValue, false)
}
```

Reply:
Or what about doing `storage = window?.localStorage` as the last argument? Even JS alone seems to allow optional chaining on default value assignment, so that should avoid the dreaded "can't access property `localStorage` of undefined" error in the server environment.

Reply:
Huh, I guess I don't really understand serverside rendering with react ... I thought this wouldn't work but it does!

Reply:
`window?...` never works for some reason - that's why `typeof window === "undefined" ? ...` is sprinkled everywhere when the optional chaining would be simpler

Reply:
I think it doesn't work because `?.` doesn't let you test something that hasn't been declared

I made another attempt at doing this refactor in https://github.com/Couchers-org/couchers/pull/2669/commits/2ea0a364698272294cf552fdbb531df8315bc804. I stole some "inspiration" from here https://dev.to/devlargs/nextjs-hook-for-accessing-local-or-session-storage-variables-3n0

Also I see there's a way to bypass ssr with next, but not sure if that's ideal or not https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr

Reply:
I wonder if we can do `globalThis?.window` based on this MDN page about `[globalThis` being the standard way to access the global context/this object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis)